### PR TITLE
Change AWS terraform configuration to work with Terraform 0.13+

### LINKED ISSUE #196: Update terraform files to support terraform >0.13
`terraform init` in `deploy/terraform-aws` fails with 
```
There are some problems with the configuration, described below.

The Terraform configuration must be valid before initialization so that
Terraform can determine which modules and providers need to be installed.

Warning: Version constraints inside provider configuration blocks are deprecated

  on main.tf line 6, in provider "external":
   6:   version = "~> 1.2"

Terraform 0.13 and earlier allowed provider version constraints inside the
provider configuration block, but that is now deprecated and will be removed
in a future version of Terraform. To silence this warning, move the provider
version constraint into the required_providers block.

(and 4 more similar warnings elsewhere)


Warning: Interpolation-only expressions are deprecated

  on main.tf line 28, in locals:
  28:     "Name"                                      = "${var.cluster_name}"

Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.

Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.

(and one more similar warning elsewhere)


Error: Invalid reference from destroy provisioner

  on main.tf line 57, in resource "aws_vpc" "main":
  57:     command     = "${path.module}/cleanup-vpc.sh ${self.id} ${var.cluster_name}"

Destroy-time provisioners and their connection configurations may only
reference attributes of the related resource, via 'self', 'count.index', or
'each.key'.

References to other resources during the destroy phase can cause dependency
cycles and interact poorly with create_before_destroy.


Error: Invalid reference from destroy provisioner

  on main.tf line 59, in resource "aws_vpc" "main":
  59:     environment = {
  60:       "AWS_REGION"         = var.region
  61:       "AWS_DEFAULT_REGION" = var.region
  62:     }

Destroy-time provisioners and their connection configurations may only
reference attributes of the related resource, via 'self', 'count.index', or
'each.key'.

References to other resources during the destroy phase can cause dependency
cycles and interact poorly with create_before_destroy.


Error: Invalid reference from destroy provisioner

  on main.tf line 59, in resource "aws_vpc" "main":
  59:     environment = {
  60:       "AWS_REGION"         = var.region
  61:       "AWS_DEFAULT_REGION" = var.region
  62:     }

Destroy-time provisioners and their connection configurations may only
reference attributes of the related resource, via 'self', 'count.index', or
'each.key'.

References to other resources during the destroy phase can cause dependency
cycles and interact poorly with create_before_destroy.


Error: Invalid reference from destroy provisioner

  on main.tf line 73, in resource "aws_internet_gateway" "gw":
  73:     command     = "${path.module}/cleanup-vpc.sh ${self.vpc_id} ${var.cluster_name}"

Destroy-time provisioners and their connection configurations may only
reference attributes of the related resource, via 'self', 'count.index', or
'each.key'.

References to other resources during the destroy phase can cause dependency
cycles and interact poorly with create_before_destroy.


Error: Invalid reference from destroy provisioner

  on main.tf line 75, in resource "aws_internet_gateway" "gw":
  75:     environment = {
  76:       "AWS_REGION"         = var.region
  77:       "AWS_DEFAULT_REGION" = var.region
  78:     }

Destroy-time provisioners and their connection configurations may only
reference attributes of the related resource, via 'self', 'count.index', or
'each.key'.

References to other resources during the destroy phase can cause dependency
cycles and interact poorly with create_before_destroy.


Error: Invalid reference from destroy provisioner

  on main.tf line 75, in resource "aws_internet_gateway" "gw":
  75:     environment = {
  76:       "AWS_REGION"         = var.region
  77:       "AWS_DEFAULT_REGION" = var.region
  78:     }

Destroy-time provisioners and their connection configurations may only
reference attributes of the related resource, via 'self', 'count.index', or
'each.key'.

References to other resources during the destroy phase can cause dependency
cycles and interact poorly with create_before_destroy.
```
we should fix those errors and warnings, so they works with latest terraform version.

### PR DESCRIPTION
**Pull request by henryprecheur**
Still work in progress. I did AWS, and GCE will be next.

Closes #196

Update Terraform configuration files to support 0.13+. The maintainers have deprecated a few things like referencing external variables from a destroy provisioner.

### CODE REVIEW

**Comment by myechuri** [comment]
Congratulations on your first kip PR @henryprecheur !

**Comment by henryprecheur** [comment]
The first of many! ;-)

**Comment by henryprecheur** [comment]
I re-ran the Terraform script 3 times and everytime it managed to deleted everything. I’m not sure why the Travis jobs failed because of the VPC :-/

**Comment by jrroman** [comment]
It could possibly be that there was an eni or something of the like that was not deallocated quickly enough for the vpc to part with it.

File: deploy/terraform-aws/main.tf [hunk 1]
```diff
@@ -52,33 +54,50 @@ resource "aws_vpc" "main" {
...
+resource "null_resource" "main" {
+  triggers = {
+    cluster_name = var.cluster_name
+    region = var.region
+  }
+
   provisioner "local-exec" {
     when        = destroy
-    command     = "${path.module}/cleanup-vpc.sh ${self.id} ${var.cluster_name}"
+    command     = "./cleanup-vpc.sh ${self.id} ${self.triggers.cluster_name}"
```
Comment (ldx) [inline deploy/terraform-aws/main.tf:68]: The first parameter should be the VPC ID, here the ID of the `null_resource` is passed in.

File: deploy/terraform-aws/README.md [hunk 2]
```diff
@@ -7,15 +7,35 @@ The Terraform config here can be used to provision a simple test cluster with ki
...
+    [... lots of output ...]
+    $ terraform plan
+    [... lots of output ...]
+
+[Upload][0] your SSH public key to AWS, and remember its name. Make
+sure the private key for this public key is your default SSH key:
+`$HOME/.ssh/id_rsa` (NB: investigate the SSH key situation to see if
+there’s an easier way to get this set-up correctly.) In this example
+we’ll name our key-pair “alice-key”. Come up with a unique cluster
+name like “alice-cluster”. Finally put these values into a tfvars file:
```
[the remark this answers is not in the document]
Comment (henryprecheur) [reply deploy/terraform-aws/README.md:28]: Yeah, I haven’t figured out how to get the key from AWS work. I’ll update the document later.

File: deploy/terraform-aws/README.md [hunk 3]
```diff
@@ -7,15 +7,35 @@ The Terraform config here can be used to provision a simple test cluster with ki
...
+
+    $ cat > myenv.tfvars <<EOF
+    ssh_key_name = "alice-key"
+    cluster_name = "alice-cluster"
+    EOF
+
 You can then apply your config:
 
-    cp env.tfvars.example myenv.tfvars
-    vi myenv.tfvars # You can change settings for your cluster here.
```
Comment (ldx) [inline deploy/terraform-aws/README.md:18]: Is there a hint somewhere else in the README to use `env.tfvars.example` to create a customized `myenv.tfvars`?

**Review by ldx — APPROVED** [review]

File: deploy/terraform-aws/README.md (hunk 3 above)
Comment (henryprecheur) [reply deploy/terraform-aws/README.md:18]: I'll add something in the documentation for that.

**Pull request merged**